keps/sig-scheduling/3022-min-domains-in-pod-topology-spread/README.md

```diff
@@ -0,0 +1,468 @@
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+With [Pod Topology Spread](/keps/sig-scheduling/895-pod-topology-spread), users can define the rule to spread pods across your cluster among failure-domains.
+And, we propose to add the parameter `minDomains` to limit the minimum number of domains in Pod Topology Spread.
+
+## Motivation
+
+Pod Topology Spread has [`maxSkew` parameter](https://github.com/kubernetes/enhancements/tree/11a976c74e1358efccf251d4c7611d05ce27feb3/keps/sig-scheduling/895-pod-topology-spread#maxskew), which control the degree to which Pods may be unevenly distributed.
+But, there isn't a way to control the number of domains over which we should spread.
+
+### Goals
+
+Users can define a minimum number of domains with `minDomains` parameter.
+
+When the number of domains that have matching Pods is less than `minDomains`,
+
+- when `whenUnsatisfiable` equals `DoNotSchedule`, scheduler doesn't schedule a matching Pod to Nodes on the domains that have matching Pods.
+  i.e. scheduler filters Nodes on the domains that have matching Pods in the `Filter` phase.
+- when `whenUnsatisfiable` equals `ScheduleAnyway`, scheduler prefers Nodes on the domains that don't have matching Pods.
+  i.e. scheduler gives a high score to Nodes on the domains that don't have matching Pods in the `Score` phase.
+
+## Proposal
+
+### User Story
+
+I am using cluster autoscaler and I want to force spreading a deployment over at least 10 Nodes.
+
+## Design Details
+
+### API
+
+New parameter called `MinDomains` is introduced.
+
+```go
+type TopologySpreadConstraint struct {
+......
+  // MinDomains describes the minimum number of domains.
+  // When the number of domains that have matching Pods is less than `minDomains`,
+  // - when `whenUnsatisfiable` equals `DoNotSchedule`, scheduler doesn't schedule a matching Pod to Nodes on the domains that have matching Pods.
+  //   i.e. scheduler filters Nodes on the domains that have matching Pods in the `Filter` phase.
+  // - when `whenUnsatisfiable` equals `ScheduleAnyway`, scheduler prefers Nodes on the domains that don't have matching Pods.
+  //   i.e. scheduler gives a high score to Nodes on the domains that don't have matching Pods in the `Score` phase.
+  // +optional
+  MinDomains int32
+}
+```
+
+### How user stories are addressed
+
+Users can set `MinDomains` and `whenUnsatisfiable: DoNotSchedule` to achieve it.
+
+```yaml
+spec:
```

Comment:
Can you give a full Deployment example to show the replicas?

Reply:
Sure.